soc/riscv/riscv-privileged/common/soc_common_irq.c

```diff
@@ -38,16 +41,26 @@ void z_riscv_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flag
 
 #else /* PLIC + HLINT/CLINT or HLINT/CLINT only */
 
+#if DT_HAS_COMPAT_STATUS_OKAY(sifive_plic_1_0_0)
+#define riscv_ic_irq_enable riscv_plic_irq_enable
+#define riscv_ic_irq_disable riscv_plic_irq_disable
+#define riscv_ic_irq_is_enabled riscv_plic_irq_is_enabled
+#elif DT_HAS_COMPAT_STATUS_OKAY(scr_ipic)
+#define riscv_ic_irq_enable scr_ipic_irq_enable
+#define riscv_ic_irq_disable scr_ipic_irq_disable
+#define riscv_ic_irq_is_enabled scr_ipic_irq_is_enabled
+#endif
```

Comment:
This is not going to work because it is not scalable (what if we have to add one more interrupt controller by some other vendor in the future?).

Is this IPIC a new controller derived from PLIC? In that case your driver must define `riscv_plic_irq_*` functions. If this is not PLIC-compliant and it is something totally new / different then you must separate the PLIC code-path from yours in this file, entirely.

Reply:
Hi @carlocaione , thanks for your feedback on this.

Our IC is indeed different to PLIC. And originally it was implemented the way you've mentioned - we just duplicated the PLIC code path here and renamed e.g riscv_plic_irq_enable to scr_ipic_irq_enable. 

But that felt like unjustified duplication since the (PLIC) functionality in soc_common_irq.c looks generic to any 2-level IC, and the only difference is the actual implementation of _irq_enable, _irq_disable and _irq_is_enabled, which could be instantiated by the particular IC.

So here we made an attempt to unify the approach for different ICs without changing the PLIC code. 
In theory we can introduce something like riscv_irq_enable/disable/is_enabled RISC-V wide API to be implemented by particular IC (including PLIC) and turn soc_common_irq.c to kind of a glue layer between the platform-independent IRQ infrastructure and particular RISC-V interrupt controllers implementation.

What do you think?

Reply:
Is your IC following any spec or is a totally custom interrupt controller? Because right now we only have `CONFIG_RISCV_HAS_CLIC` and `CONFIG_RISCV_HAS_PLIC` so here you are implicitly assuming that your interrupt controller is a PLIC.

> In theory we can introduce something like riscv_irq_enable/disable/is_enabled RISC-V wide API to be implemented by particular IC (including PLIC) and turn soc_common_irq.c to kind of a glue layer between the platform-independent IRQ infrastructure and particular RISC-V interrupt controllers implementation.

Yes, this is the best way forward IMO. But you also must add a new `CONFIG_RISCV_HAS_...` if you are dealing with something different than CLIC and PLIC.

Reply:
@carlocaione , 

I believe IPIC is a custom one. Here is a description I've found in the net, it's a part of the SCR1 platform we plan to upstream:
https://caxapa.ru/thumbs/830519/scr1_eas.pdf

Is it ok, if we use CONFIG_RISCV_HAS_IPIC with custom IC?

If we proceed down the riscv_irq_enable/disable/is_enabled path, how would we design this? Should we introduce a new config, say CONFIG_RISCV_UNIFIED_IRQ_API_SUPPORT (which will be selected for PLIC, IPIC, etc) or something?

Then we could have e.g.

```
void arch_irq_enable(unsigned int irq)
{
	uint32_t mie;

#if defined(CONFIG_RISCV_UNIFIED_IRQ_API_SUPPORT)
	unsigned int level = irq_get_level(irq);

	if (level == 2) {
		irq = irq_from_level_2(irq);
		riscv_irq_enable(irq);
		return;
	}
#endif
	/*
	 * CSR mie register is updated using atomic instruction csrrs
	 * (atomic read and set bits in CSR register)
	 */
	mie = csr_read_set(mie, 1 << irq);
}
```

or we may use some form of it, e.g.

```
void arch_irq_enable(unsigned int irq)
{
	uint32_t mie;

#if defined(riscv_irq_enable)
unsigned int level = irq_get_level(irq);

	if (level == 2) {
		irq = irq_from_level_2(irq);
		riscv_irq_enable(irq);
		return;
	}
#endif

.... 
```
What do you think?

Reply:
> Is it ok, if we use CONFIG_RISCV_HAS_IPIC with custom IC?

If `IPIC` is not a formal standard I would not go that way. Instead I would add something like we have in ARM for `CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER` that can be used by custom interrupt controllers. Please check how that is done for ARM.

Reply:
Hi @carlocaione , thanks!

Just to clarify that I've understood you correctly:
1.) we introduce CONFIG_RISCV_CUSTOM_INTERRUPT_CONTROLLER, which is selected when IPIC is enabled
2.) we define a set of functions like riscv_irq_enable/disable/is_enabled which is to be implemented by custom IC
3.) we duplicate the PLIC path in soc_common_irq.c (and we don't optimize existing PLIC logic in any way) and have it e.g. like this:

```
#if defined(CONFIG_RISCV_CUSTOM_INTERRUPT_CONTROLLER)
void arch_irq_enable(unsigned int irq)
{
	uint32_t mie;

	unsigned int level = irq_get_level(irq);
#if defined(CONFIG_2ND_LEVEL_INTERRUPTS)
	if (level == 2) {
		irq = irq_from_level_2(irq);
		riscv_irq_enable(irq);
		return;
	}
#endif
	/*
	 * CSR mie register is updated using atomic instruction csrrs
	 * (atomic read and set bits in CSR register)
	 */
	mie = csr_read_set(mie, 1 << irq);
}

....

#endif
```

Is it correct?

Reply:
Thanks! Then I suggest we proceed the second way to avoid code duplication.
@carlocaione do we need a separate commit for this within this PR or do we need entirely new PR?

Reply:
BTW https://github.com/zephyrproject-rtos/zephyr/pull/62877